projects/http-auth/main.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"encoding/base64"
+	"fmt"
+	"html"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	auth "github.com/abbot/go-http-auth"
+	"github.com/joho/godotenv"
+	"golang.org/x/time/rate"
+)
+
+// Get    curl -i 'http://localhost:8080?foo=<strong>bar</strong>'
+// Post:  curl -i -d "<em>Hi</em>" 'http://localhost:8080?foo=<strong>bar</strong>'
+// admin:long-memorable-password
+// Auth1:  curl -i 'http://localhost:8080/authenticated1' -H 'Authorization: Basic YWRtaW46bG9uZy1tZW1vcmFibGUtcGFzc3dvcmQ='
+// Auth2:  curl -i 'http://localhost:8080/authenticated2' -H 'Authorization: Basic YWRtaW46bG9uZy1tZW1vcmFibGUtcGFzc3dvcmQ='
+
+func main() {
+	if err := godotenv.Load(); err != nil {
+		fmt.Fprintf(os.Stderr, "can't read the env: %s", err)
+	}
+	http.HandleFunc("/", getHandler)
+	http.HandleFunc("/404", http.NotFound)
+	http.HandleFunc("/500", internalServerError)
+	http.HandleFunc("/authenticated1", authenticateHandler)
+
+	authenticator := auth.NewBasicAuthenticator("localhost:8080", secret)
+	http.HandleFunc("/authenticated2", authenticator.Wrap(authHandler))
+
+	http.HandleFunc("/limited", limitHandler)
+
+	fmt.Fprintf(os.Stdout, "Listening on port 8080")
+	if err := http.ListenAndServe(":8080", nil); err != nil {
+		fmt.Fprintf(os.Stderr, "Port is not empty %s", err)
```

Comment:
Is this definitely the thing that went wrong? If not, I would change the message - you don't want to mislead people. Maybe something like "Couldn't listen on port - maybe it's not free"? To suggest your level of confidence in the problem?